### ISSUE TITLE
Bit-sized integers allow strange syntax

### ISSUE THREAD

**Issue by Mingun** [kaitai-io/kaitai_struct#939]
This definitions is valid and it seems they should be forbidden:
```yaml
seq:
  - id: zeroes
    type: b00000008 # == b8
  - id: zero
    type: b0 # should be forbidden
```

**Comment by KOLANICH** [comment]
No, it doesn't seem they must be forbidden.

**Comment by generalmimon** [comment]
I'm with @KOLANICH here, I think that the first two are totally OK.

Some people may prefer padding numbers to the right to achieve some look, for example:

```ksy
seq:
  - type: b01
  - type: b53
```

I don't see anything wrong with that, and I don't think that forbidding it brings something positive.

---

The following

```ksy
  - id: zero
    type: b0
```

is also legitimate. We allow `size: 0`, which may also be useful (as well as `repeat-expr: 0`, etc.), so why not `type: b0`.

One use that comes to my mind is when you want to force the compiler to insert an `_io.alignToByte()` call without actually parsing anything - that could look like this:

```ksy
 - type: b0
 - size: 0
```

Sure, it's a bit of a hack, but a potentially _useful_ one :)

---

This situation

 ```ksy
  - id: unnecessary_be
    type: b1be
  - id: unnecessary_le
    type: b1le
```

is mentioned in the User Guide (https://doc.kaitai.io/user_guide.html#bit-endian):

> _Big-endian_ and _little-endian_ bit integers can follow **only on a byte boundary**. They can’t share the same byte. Joining them on an unaligned bit position is _undefined behavior_, and future versions of KSC will throw a compile-time error if they detect such a situation.

It is definitely planned to make the compiler detect this, see https://github.com/kaitai-io/kaitai_struct/issues/155#issuecomment-718239225.

**Comment by Mingun** [comment]
> I don't see anything wrong with that, and I don't think that forbidding it brings something positive

That seems valuable, but it is inconsistent with that fact that `s01`, `u0008` and so on is not a synonymous to the `s1` and `u8`. Such differences in behaviour give the impression of poor language design.

> We allow `size: 0`, which may also be useful (as well as `repeat-expr: 0`, etc.), so why not `type: b0`.

I believe that it is only because nobody thought about that corner cases (what is the fact that _negative_ values are also not prohibited). In my opinion we should investigate possible usages of that strange definitions and forbid them if nothing will be found. If they can be used for some workarounds it is better to investigate what the workaround tries to do and introduce a dedicated syntax for that.

> It is definitely planned to make the compiler detect this, 

This example is not about two fields following each other, they put to the one type only as an example of definitions. Treat the type as containing only one of the specified fields. What the means to define a bit-order of one-bit field?

**Comment by generalmimon** [comment]
@Mingun:

> What the means to define a bit-order of one-bit field?

I highly recommend reading https://doc.kaitai.io/user_guide.html#_bit_sized_integers.

Note that in the context of Kaitai Struct, the term `bit-endian` dictates the "parsing start and direction" (in this case, the `be`/`le` suffix of `b1` affects the position where it shall parsed within the enclosing byte in the stream - if it starts on the stream byte's **most** significant bit (`bit-endian: be`) or the **least** significant bit (`bit-endian: le`)).

It **does not** mean [bit numbering](https://en.wikipedia.org/wiki/Bit_numbering) of the parsed bits (in fact, that is constant for both `bit-endian: be/le` - notice that on all diagrams in [Bit-sized integers (User Guide)](https://doc.kaitai.io/user_guide.html#_bit_sized_integers), the integer bits are always in order `a4`, `a3`, ..., `a0`, regardless of whether big/little-endian parsing direction is used).

tl;dr:

<pre><code>seq:
  - id: p
    type: b1<ins><strong>be</strong></ins></code></pre>

```
             d[0]
7   6   5   4   3   2   1   0
p0
┬─
p
```

> _Examples:_
> (a)
> 
> <pre><code>Offset (h)  00        01        ...
> 
>                8e<sub>16</sub>
> 00000000    <ins><strong>1</strong></ins>0001110<sub>2</sub></code></pre>
> 
> <pre><code>└─ p = <strong>true</strong></code></pre>
> ---
>
> (b)
> <pre><code>Offset (h)  00        01        ...
> 
>                7f<sub>16</sub>
> 00000000    <ins><strong>0</strong></ins>1111111<sub>2</sub></code></pre>
> 
> <pre><code>└─ p = <strong>false</strong></code></pre>
>

---

<pre><code>seq:
  - id: q
    type: b1<ins><strong>le</strong></ins></code></pre>

```
             d[0]
7   6   5   4   3   2   1   0
                            q0
                            ┬─
                            q
```


> _Examples:_
> (a)
>
> <pre><code>Offset (h)  00        01        ...
> 
>                8e<sub>16</sub>
> 00000000    1000111<ins><strong>0</strong></ins><sub>2</sub></code></pre>
> 
> <pre><code>└─ q = <strong>false</strong></code></pre>
> ---
>
> (b)
> <pre><code>Offset (h)  00        01        ...
> 
>                01<sub>16</sub>
> 00000000    0000000<ins><strong>1</strong></ins><sub>2</sub></code></pre>
> 
> <pre><code>└─ p = <strong>true</strong></code></pre>
>
